### PR TITLE
HIVE-27195:Add database authorization for drop table command

### PR DESCRIPTION
**Pull request by rtrivedi12**
This change includes authorization of the database object during the "drop table" command. Similar to "Create table", DB permissions should be verified in the case of "drop table" too. Also, In case of a temporary table drop, empty input, and output HivePrivilegeObject are sent to the authorizer as temporary tables are skipped from authorization.

### What changes were proposed in this pull request?
Authorize write actions on the database during drop table action, and add the database object to the list of output objects sent for verifying privileges.

### Why are the changes needed?
To prevent unauthorized users from dropping temp tables from any database

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Running existing qtests and adding new tests.

### CODE REVIEW

**Comment by rtrivedi12** [comment]
@saihemanth-cloudera Can you please review

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java [hunk 1]
```diff
@@ -46,9 +48,13 @@ public DropTableAnalyzer(QueryState queryState) throws SemanticException {
 
   @Override
   public void analyzeInternal(ASTNode root) throws SemanticException {
-    String tableName = getUnescapedName((ASTNode) root.getChild(0));
+    TableName qualTabName = getQualifiedTableName((ASTNode) root.getChild(0));
+    String tableName = qualTabName.getTable();
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java:52]: `qualTabName.getTable()` -> `qualTabName.getNotEmptyDbTable`

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java [hunk 2]
```diff
@@ -46,9 +48,13 @@ public DropTableAnalyzer(QueryState queryState) throws SemanticException {
...
     boolean ifExists = (root.getFirstChildWithType(HiveParser.TOK_IFEXISTS) != null);
     boolean throwException = !ifExists && !HiveConf.getBoolVar(conf, ConfVars.DROP_IGNORES_NON_EXISTENT);
+    //Authorize database for drop table command
+    Database database  = getDatabase(qualTabName.getDb());
+    outputs.add(new WriteEntity(database, WriteType.DDL_NO_LOCK));
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java:57]: I find when we create table, the `WriteType` is `DDL_SHARED`, can we change it to that?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java (hunk 1 above)
Comment (rtrivedi12) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java:52]: Thanks for reviewing! Incorporated the change.

**Comment by dengzhhu653** [comment]
There are some qfile tests failed, I think it due to we add the database to the output, cloud you please fix these as well? thank you!

**Review by dengzhhu653 — APPROVED** [review]
nit: you can restore the TestPigHBaseStorageHandler, TestDDLWithRemoteMetastoreSecondNamenode, TestDatabaseTableDefault, TestTxnExIm, TestHiveStrictManagedMigration, these tests can run fine with the latest DropTableAnalyzer, others LGTM.
Thank you for the work!

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 3]
```diff
@@ -140,7 +140,7 @@ private void dropTables() throws Exception {
     public void tearDown() throws Exception {
         try {
             if (d != null) {
-                dropTables();
+                //dropTables();
```
Comment (zabetak) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java:143]: Why is this change needed?

If it is really needed then the code should be removed; it is generally best to avoid leaving code commented out.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java [hunk 4]
```diff
@@ -46,9 +48,16 @@ public DropTableAnalyzer(QueryState queryState) throws SemanticException {
...
   @Override
   public void analyzeInternal(ASTNode root) throws SemanticException {
-    String tableName = getUnescapedName((ASTNode) root.getChild(0));
+    TableName qualTabName = getQualifiedTableName((ASTNode) root.getChild(0));
+    String tableName = qualTabName.getNotEmptyDbTable();
     boolean ifExists = (root.getFirstChildWithType(HiveParser.TOK_IFEXISTS) != null);
     boolean throwException = !ifExists && !HiveConf.getBoolVar(conf, ConfVars.DROP_IGNORES_NON_EXISTENT);
+    //Authorize database for drop table command
+    // Skip db object if database doesn't exist
+    Database database  = getDatabase(qualTabName.getDb(),false);
```
[the remark this answers is not in the document]
Comment (rtrivedi12) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java:57]: @zabetak  In case of the DROP TABLE IF EXISTS command, this `throwException` refers to throwing exception in case of a non-existent table during the `getTable` parameter. As we want NOOP in case of non-existent DB in DROP TABLE, hardcoded this as  false for `getDatabase` call

File: ql/src/test/org/apache/hadoop/hive/ql/TestTxnExIm.java [hunk 5]
```diff
@@ -62,11 +62,12 @@ public void testExportCustomDb() throws Exception {
...
     int[][] rows1 = {{1, 2}, {3, 4}};
     final String tblName = useDefualtDb ? "T" : "foo.T";
-    runStatementOnDriver("drop table if exists " + tblName);
-    runStatementOnDriver("drop table if exists TImport ");
     if(!useDefualtDb) {
       runStatementOnDriver("create database foo");
     }
+    runStatementOnDriver("drop table if exists " + tblName);
+    runStatementOnDriver("drop table if exists TImport ");
+
```
[the remark this answers is not in the document]
Comment (rtrivedi12) [reply ql/src/test/org/apache/hadoop/hive/ql/TestTxnExIm.java:70]: Will revert these tests.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java (hunk 3 above)
Comment (rtrivedi12) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java:143]: A few tests using drop table statements before creating the database resulted in 'Database does not exist' error with the last patch. Hence used DROP DATABASE CASCADE instead of dropping the table first and dropping the DB. Anyways, as we decided to silently return OK in case of non-existent DB, will revert these tests.

**Comment by rtrivedi12** [comment]
Thanks for the review @dengzhhu653 and @zabetak!

**Review by zabetak — COMMENTED** [review]
The PR/JIRA ticket reports that there is an error when using `DROP TABLE IF EXISTS` but there were no test failures before this change and there are no new tests added as part of the changes here which indicates that there may not be enough coverage.

Test cases are also an implicit documentation of how things are supposed to work so it would be great if we could add a few. Is there any particular reason of why the PR does not add tests covering/reproducing the reported error?

**Comment by rtrivedi12** [comment]
> The PR/JIRA ticket reports that there is an error when using `DROP TABLE IF EXISTS` but there were no test failures before this change and there are no new tests added as part of the changes here which indicates that there may not be enough coverage.
> 
> Test cases are also an implicit documentation of how things are supposed to work so it would be great if we could add a few. Is there any particular reason of why the PR does not add tests covering/reproducing the reported error?

This error/behavior change with `drop table if exists` is observed only with Ranger for unauthorized user( After RANGER-3407). Hence, there were no test failures. I had a discussion with the Hive QE team and they mentioned that this is a gap with QE testing too as all the variations with `drop table if Exists` are run with full permissions on Ranger.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java [hunk 6]
```diff
@@ -46,9 +48,16 @@ public DropTableAnalyzer(QueryState queryState) throws SemanticException {
...
+    if (database != null) {
+      outputs.add(new WriteEntity(database, WriteType.DDL_SHARED));
```
Comment (zhangbutao) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java:59]: Should we be consistent with HMS server side Ranger auth?
https://github.com/apache/hive/blob/3bc62cbc2d42c22dfd55f78ad7b41ec84a71380f/ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/events/DropTableEvent.java#L52C129-L52C129

**Review by zabetak — COMMENTED** [review]
@rtrivedi12 Many thanks for your hard work. I think we may need to add a few more test cases around this feature. I left a comment about this under the [JIRA](https://issues.apache.org/jira/browse/HIVE-27195?focusedCommentId=17744625&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17744625).

File: ql/src/test/queries/clientpositive/authorization_drop_table.q [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+show grant user hive_test_user on table drop_table_auth_1;
+
+CREATE TEMPORARY TABLE drop_temp_table LIKE drop_table_auth_1;
+
+set hive.security.authorization.enabled=true;
+
+-- Drop table works fine as user has privs for both DB and table
+drop table if exists drop_table_auth_1;
+
+-- Dropping temporary table does not require authorization
```
[the remark this answers is not in the document]
Comment (rtrivedi12) [reply ql/src/test/queries/clientpositive/authorization_drop_table.q:23]: I think the comment is misleading. I meant dropping temp does not require table privileges but we still need database privileges. 

`REVOKE DROP ON DATABASE auth_drop_table TO USER hive_test_user;
DROP TABLE IF EXISTS drop_temp_table;`

This would fail.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java (hunk 2 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java]: note, in AbstractDropPartitionAnalyzer we only add outputs for partitions.

File: ql/src/test/queries/clientnegative/authorization_droptable_fail_10.q [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+set hive.security.authorization.enabled=false;
+
+CREATE DATABASE auth_db_fail_4;
+use auth_db_fail_4;
+create temporary table drop_table_temp_4 (key int, value string) partitioned by (ds string);
+
+-- Drop temporary table from current database with IF EXISTS
+
+set hive.security.authorization.enabled=true;
+DROP TABLE IF EXISTS drop_table_temp_4;
```
Comment (zabetak) [inline ql/src/test/queries/clientnegative/authorization_droptable_fail_10.q:11]: nit: Some files miss the line ending.

File: ql/src/test/queries/clientnegative/authorization_droptable_fail_3.q [hunk 9]
```diff
@@ -0,0 +1,14 @@
+
+
```
Comment (zabetak) [inline ql/src/test/queries/clientnegative/authorization_droptable_fail_3.q:2]: nit: useless blank lines

**Review by zabetak — COMMENTED** [review]
The tests look good I have a few questions and some nit comments. Don't bother addressing the nit comments unless it is necessary to commit more changes to the PR.

File: ql/src/test/queries/clientpositive/authorization_drop_table.q [hunk 10]
```diff
@@ -1,25 +1,96 @@
...
+create table drop_table_auth_3 (key int, value string) partitioned by (ds string);
+create table drop_table_auth_4 (key int, value string);
+CREATE TEMPORARY TABLE auth_temp_table_1(key STRING, c1 INT, c2 STRING) STORED AS TEXTFILE;
+CREATE TEMPORARY TABLE auth_temp_table_2(key STRING, c1 INT, c2 STRING) STORED AS TEXTFILE;
 
-show grant user hive_test_user on table drop_table_auth_1;
 
-CREATE TEMPORARY TABLE drop_temp_table LIKE drop_table_auth_1;
+GRANT All on table auth_db_1.drop_table_auth_4 to user hive_test_user;
+GRANT All on table auth_db_1.drop_table_auth_3 to user hive_test_user;
```
[the remark this answers is not in the document]
Comment (rtrivedi12) [reply ql/src/test/queries/clientpositive/authorization_drop_table.q]: Yeah, That's right! `GRANT DROP ON DATABASE` is sufficient for dropping the tables. Updated tests. For temporary tables, since the table object is skipped from being added to Hive Privileges I/O objects. Hence, no privileges are required for temporary tables.

**Review by zabetak — APPROVED** [review]
LGTM pending tests, conflicts, and minor comments. When these are addressed and assuming there are no new blocking comments from other reviewers we can merge the PR in 24h.

File: ql/src/test/queries/clientnegative/authorization_droptable_fail_6.q [hunk 11]
```diff
@@ -0,0 +1,9 @@
+set hive.security.authorization.manager=org.apache.hadoop.hive.ql.security.authorization.DefaultHiveAuthorizationProvider;
+set hive.security.authorization.enabled=false;
+
+CREATE TEMPORARY TABLE auth_temp_table_2(key STRING, c1 INT, c2 STRING) STORED AS TEXTFILE;
+
+-- Drop temporary table with IF EXISTS WITHOUT DB Drop Privileges
+
+set hive.security.authorization.enabled=true;
+DROP TABLE auth_temp_table_2;
```
[the remark this answers is not in the document]
Comment (rtrivedi12) [reply ql/src/test/queries/clientnegative/authorization_droptable_fail_6.q:9]: I think you meant `authorization_droptable_fail_5.q` and `authorization_droptable_fail_6.q` are identical. Thanks for the catch! The only difference between these two is of `IF EXISTS` clause. Corrected the test.

File: ql/src/test/queries/clientnegative/authorization_droptable_fail_11.q [hunk 12]
```diff
@@ -0,0 +1,12 @@
+set hive.exec.drop.ignorenonexistent=false;
+
+set hive.security.authorization.manager=org.apache.hadoop.hive.ql.security.authorization.DefaultHiveAuthorizationProvider;
+
+-- Drop non-existing table WITHOUT DB Drop Privileges
```
[the remark this answers is not in the document]
Comment (rtrivedi12) [reply ql/src/test/queries/clientnegative/authorization_droptable_fail_11.q:5]: Updated!
